File: backend/app/agents/supervisor.py
"""Supervisor Agent.

Routes incoming events to the specialist agent responsible for them and
returns that specialist's result. This is the single dispatch point used
by the event system (Phase 5).
"""

from typing import Any

from sqlalchemy.orm import Session

from app.agents.base import AgentResult, BaseAgent
from app.agents.climate_agent import ClimateRiskAgent
from app.agents.customs_agent import CustomsAgent
from app.agents.demand_agent import DemandIntelligenceAgent
from app.agents.logistics_agent import LogisticsAgent


class SupervisorAgent(BaseAgent):
    name = "supervisor"

    def __init__(self) -> None:
        super().__init__()
        self.demand = DemandIntelligenceAgent()
        self.logistics = LogisticsAgent()
        self.climate = ClimateRiskAgent()
        self.customs = CustomsAgent()

        # Event → specialist routing table
        self.routes: dict[str, BaseAgent] = {
            "buyer.request.created": self.demand,
            "crop.harvest.ready": self.demand,
            "weather.alert": self.climate,
            "shipment.delayed": self.logistics,
            "shipment.departed": self.customs,
            "customs.approved": self.logistics,
        }

    def handle(self, db: Session, payload: dict[str, Any]) -> AgentResult:
        event = payload.get("event", "")
        specialist = self.routes.get(event)
        if specialist is None:
            return AgentResult(
                agent=self.name,
                action="no_route",
                confidence=1.0,
                rationale=f"No specialist registered for event {event!r}",
                outputs={"event": event},
            )

        result = specialist.run(db, payload)
        return AgentResult(
            agent=self.name,
            action=f"delegated:{specialist.name}",
            confidence=result.confidence,
            rationale=f"Routed {event!r} to {specialist.name}",
            outputs=result.as_dict(),
        )

    def dispatch(self, db: Session, event: str, payload: dict[str, Any]) -> AgentResult:
        """Convenience entry point for the event bus."""
        return self.run(db, {**payload, "event": event})


supervisor = SupervisorAgent()

File: src/lib/server/agents/supervisor.ts
/**
 * Supervisor Agent.
 *
 * Routes each coordination event to the specialist responsible for it and
 * returns that specialist's result. Single dispatch point for the event system.
 */

import { AgronomyAgent } from "./agronomy";
import { BaseAgent, result, type AgentPayload, type AgentResult } from "./base";
import { ClimateRiskAgent } from "./climate";
import { DemandIntelligenceAgent } from "./demand";
import { LogisticsAgent } from "./logistics";
import { PlantingCoordinationAgent } from "./planting";

export class SupervisorAgent extends BaseAgent {
  readonly name = "supervisor";

  readonly demand = new DemandIntelligenceAgent();
  readonly logistics = new LogisticsAgent();
  readonly climate = new ClimateRiskAgent();
  readonly planting = new PlantingCoordinationAgent();
  readonly agronomy = new AgronomyAgent();

  /** Event → specialist routing table */
  private readonly routes: Record<string, BaseAgent> = {
    "substitution.gap.detected": this.demand,
    "demand.review.requested": this.demand,
    "climate.risk.elevated": this.climate,
    "storm.alert": this.climate,
    "lane.assessment.requested": this.logistics,
    "planting.window.review": this.planting,
    "soil.assessment.requested": this.agronomy,
  };

  protected async handle(payload: AgentPayload): Promise<AgentResult> {
    const event = (payload.event as string | undefined) ?? "";
    const specialist = this.routes[event];

    if (specialist === undefined) {
      return result(
        this.name,
        "no_route",
        1.0,
        `No specialist registered for event '${event}'`,
        { event }
      );
    }

    const specialistResult = await specialist.run(payload);
    return result(
      this.name,
      `delegated:${specialist.name}`,
      specialistResult.confidence,
      `Routed '${event}' to ${specialist.name}`,
      { ...specialistResult }
    );
  }

  /** Convenience entry point for the event bus. */
  dispatch(event: string, payload: AgentPayload): Promise<AgentResult> {
    return this.run({ ...payload, event });
  }
}

const globalSupervisor = globalThis as typeof globalThis & {
  __nexusGridSupervisor?: SupervisorAgent;
};

if (!globalSupervisor.__nexusGridSupervisor) {
  globalSupervisor.__nexusGridSupervisor = new SupervisorAgent();
}

/** The process-wide supervisor, so agent memory survives hot reloads. */
export const supervisor: SupervisorAgent = globalSupervisor.__nexusGridSupervisor;
